src/transaction/locator_sr.c

```diff
@@ -13879,7 +13879,35 @@ locator_multi_insert_force (THREAD_ENTRY * thread_p, HFID * hfid, OID * class_oi
   // Log the postpone operation
   heap_log_postpone_heap_append_pages (thread_p, hfid, class_oid, heap_pages_array);
 
-  return NO_ERROR;
+  OR_CLASSREP *classrepr = NULL;
+  int idx, classrepr_cacheindex = -1;
+
+  classrepr = heap_classrepr_get (thread_p, class_oid, NULL, NULL_REPRID, &classrepr_cacheindex);
+  assert (classrepr != NULL);
+
+  for (idx = classrepr->n_indexes - 1; idx >= 0; idx--)
+    {
+      if (btree_is_unique_type (classrepr->indexes[idx].type))
+	{
+	  GLOBAL_UNIQUE_STATS *stats = NULL;
+
+	  logtb_update_global_unique_stats_by_delta (thread_p, &classrepr->indexes[idx].btid,
+		recdes_array.size (), 0, recdes_array.size (), true);	
+	  stats = logtb_get_global_unique_stats_entry (thread_p, &classrepr->indexes[idx].btid, true);
+	  if (stats != NULL)
+	    {
+	      error_code = btree_reflect_global_unique_statistics (thread_p, stats, false);
+	      pthread_mutex_unlock (&stats->mutex);
+	    }
+	}
+    }
```

Comment:
From what I've heard and read, (1) all stats are kept on a memory cache, (2) the global cache is updated when a transaction commits and a log is left for this, (3) the global stats are reflected to root headers of btrees during checkpoint.

Let me ask some questions:
1. After the bulk update here, the transaction will commit. Why we should update global stats manually here? Won't the transaction including this operation commit?
2. If we leave a log during updating global stats, it seems that we don't need to reflect the state to btree root headers. Why do we have to reflect it here manually? From a jira issue, they remove reflecting during commit becauce of performance, but it is reflected for every BU here.
3. What happenes if a transaction aborts after reflecting by an internal error or something? btree_reflect_global_unique_statistics() does not leave log for UNDO.

Reply:
multi_insert_force routine is called only by loaddb command. The log mechanism is not performed while the indexes are loading. So, we should reflect the statistics of the btree header inforcely.

Reply:
In common, btree header statistics are reflected by the following mechanism, but in the case of loaddb, this mechanism does not work.

**log_commit > log_complete > logpb_checkpoint > logtb_reflect_global_unique_stats_to_btree**

So, to reflect the stats, I called the **logtb_reflect_global_unique_stats_to_btree** routine directly.
From the log manager's point of view, please comment on how it should be reflected.

Reply:
**log_commit -> logtb_complete_mvcc->logtb_update_global_unique_stats_by_delta** : leaves a log for updating the global stat without reflecting to a corresponding btree root header.
**logpb_checkpoint->logtb_reflect_global_unique_stats_to_btree**: reflects stats to the global stat to the btree root header without a log.
And if the logpb_checkpoint() is not executed during running, the global stats in the tree root header are reflected during recovery. That's what I understand. The way see it, if operations commits in loaddb and the local stats of each transactions are updated properly during the buik insert, it will work. 

1. Is there something I misunderstand? What do you mean by "does not work in the case of loaddb"?
2. Where are local stats updated, by the way?

Reply:
I checked the routine related to index header statistics log more detaily. **logtb_update_global_unique_stats_by_delta** already write undo/redo log, it's called from loaddb's  locator_multi_insert_force. Therefore, the log would be written correctly. please check the routine **logtb_update_global_unique_stats_by_delta**.

Reply:
Yes, that is what I mean. So, the log for reflecting log is written, then why we should reflect it manually?

Reply:
loaddb doesn't use mvcc and the loaddb transaction doesn't reach to **complete_mvcc** which is used at only mvcc.
whereas in SA_MODE **logtb_tran_update_all_global_unique_stats** is called. So, I'm goning to check if I coud modify the code to call the **logtb_tran_update_all_global_unique_stats** in SERVER mode.

Reply:
I found a bug-like at the logtb_complete_mvcc routine as below:
```
#if defined(SA_MODE)
      if (committed && logtb_tran_update_all_global_unique_stats (thread_p) != NO_ERROR)
	{
	  assert (false);
	}
#else	/* !SA_MODE */	       /* SERVER_MODE */
```
"#if defined(SA_MODE)" looks not proper, I tested the loaddb with removing this conditional macro. The statistics was reflected successfully. However, I don't know why the code is wrtten in this way.

Reply:
It seems to assume that all operations that potentially update the unique stats leave some mvcc logs. Regarding HEAP operation, the bulk updating doesn't leave MVCC logs, but I don't really know the case of index.

Reply:
I fixed the routine to update stats on SERVER_MODE too.